cpp/src/arrow/filesystem/gcsfs.cc

```diff
@@ -342,6 +344,53 @@ class GcsFileSystem::Impl {
     return internal::ToArrowStatus(CreateDirMarkerRecursive(p.bucket, p.object));
   }
 
+  Status DeleteDir(const GcsPath& p, const io::IOContext& io_context) {
+    RETURN_NOT_OK(DeleteDirContents(p, io_context));
+    if (!p.object.empty()) {
+      return internal::ToArrowStatus(client_.DeleteObject(p.bucket, p.object));
+    }
+    return internal::ToArrowStatus(client_.DeleteBucket(p.bucket));
+  }
+
+  Status DeleteDirContents(const GcsPath& p, const io::IOContext& io_context) {
+    // Deleting large directories can be fairly slow, we need to parallelize the
+    // operation. This uses `std::async()` to run multiple delete operations in parallel.
+    // A simple form of flow control limits the number of operations running in
+    // parallel.
+
+    auto async_delete =
+        [&p](gcs::Client& client,
+             google::cloud::StatusOr<gcs::ObjectMetadata> o) -> google::cloud::Status {
+      if (!o) return std::move(o).status();
+      // The list includes the directory, skip it. DeleteDir() takes care of it.
+      if (o->bucket() == p.bucket && o->name() == p.object) return {};
+      return client.DeleteObject(o->bucket(), o->name(),
+                                 gcs::Generation(o->generation()));
+    };
+
+    using Future = arrow::Future<google::cloud::Status>;
+    std::vector<Result<Future>> submitted;
+    // This iterates over all the objects, and schedules parallel deletes.
+    auto prefix = p.object.empty() ? gcs::Prefix() : gcs::Prefix(p.object);
+    for (auto& o : client_.ListObjects(p.bucket, prefix)) {
+      submitted.push_back(
```

Comment:
By using `DeferNotOk`, you could instead append to a `std::vector<Future>`, which will simplify the collection code below.

For example (untested):
```c++
    std::vector<Future> submitted;
    // This iterates over all the objects, and schedules parallel deletes.
    auto prefix = p.object.empty() ? gcs::Prefix() : gcs::Prefix(p.object);
    for (auto& o : client_.ListObjects(p.bucket, prefix)) {
      submitted.push_back(DeferNotOk(
          io_context.executor()->Submit(async_delete, std::ref(client_), std::move(o))));
    }

    return AllFinished(submitted);
```

Reply:
Neat, thanks.